src/impls.rs

```diff
@@ -0,0 +1,207 @@
+
+use crate::{ConstInit, ScopedRawMutex};
+
+#[cfg(feature = "impl-critical-section")]
+pub mod cs {
+
+    use super::*;
+
+    /// A mutex that allows borrowing data across executors and interrupts.
+    ///
+    /// # Safety
+    ///
+    /// This mutex is safe to share between different executors and interrupts.
+    pub struct CriticalSectionRawMutex {
+        taken: AtomicBool,
+    }
+    unsafe impl Send for CriticalSectionRawMutex {}
+    unsafe impl Sync for CriticalSectionRawMutex {}
+
+    impl CriticalSectionRawMutex {
+        /// Create a new `CriticalSectionRawMutex`.
+        pub const fn new() -> Self {
+            Self {
+                taken: AtomicBool::new(false),
+            }
+        }
+    }
+
+    impl ConstInit for CriticalSectionRawMutex {
+        const INIT: Self = Self::new();
+    }
+
+    unsafe impl ScopedRawMutex for CriticalSectionRawMutex {
+        #[inline]
+        #[must_use]
+        fn try_lock<R>(&self, f: impl FnOnce() -> R) -> Option<R> {
+            critical_section::with(|_| {
+                // NOTE: separated load/stores are acceptable as we are in
+                // a critical section
+                if self.taken.load(Ordering::Relaxed) {
+                    return None;
+                }
+                self.taken.store(true, Ordering::Relaxed);
+                let ret = f();
+                self.taken.store(false, Ordering::Relaxed);
+                Some(ret)
+            })
+        }
+
+        #[inline]
+        fn lock<R>(&self, f: impl FnOnce() -> R) -> R {
+            // In a critical section, it is not possible for another holder
+            // of this mutex to release, which means we have certainly
+            // reached deadlock if the lock was already locked.
+            self.try_lock(f).expect("Deadlocked")
+        }
+    }
+}
+
+// ================
+
+pub mod local {
+    use super::*;
+
+    /// A mutex that allows borrowing data in local context.
+    ///
+    /// This acts similar to a RefCell, with scoped access patterns, though
+    /// without being able to borrow the data twice.
+    pub struct LocalRawMutex {
+        taken: AtomicBool,
+        /// Prevent this from being sync or send
+        _phantom: PhantomData<*mut ()>,
+    }
+
+    impl LocalRawMutex {
+        /// Create a new `LocalRawMutex`.
+        pub const fn new() -> Self {
+            Self {
+                taken: AtomicBool::new(false),
+                _phantom: PhantomData,
+            }
+        }
+    }
+
+    unsafe impl Send for LocalRawMutex {}
+
+    impl ConstInit for LocalRawMutex {
+        const INIT: Self = Self::new();
+    }
+
+    unsafe impl ScopedRawMutex for LocalRawMutex {
+        #[inline]
+        #[must_use]
+        fn try_lock<R>(&self, f: impl FnOnce() -> R) -> Option<R> {
+            // NOTE: separated load/stores are acceptable as we are !Send and !Sync,
+            // meaning that we can only be accessed within a single thread
```

Comment:
is this specifically intended to avoid CAS operations on platforms that only have fake atomics?

Reply:
yep - if we're already paying for a critical section, no need for a second one when using portable atomic